src/main/java/bio/terra/service/configuration/ConfigurationService.java

```diff
@@ -0,0 +1,110 @@
+import org.springframework.stereotype.Component;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+@Component
+public class ConfigurationService {
+    private final Logger logger = LoggerFactory.getLogger(ConfigurationService.class);
+
+    private final SamConfiguration samConfiguration;
+
+    private Map<String, ConfigBase> configuration = new HashMap<>();
+
+    @Autowired
+    public ConfigurationService(SamConfiguration samConfiguration) {
+        this.samConfiguration = samConfiguration;
+        setConfiguration();
+    }
+
+    // -- repository API methods --
+
+    public List<ConfigModel> setConfig(ConfigGroupModel groupModel) {
+        logger.info("Setting configuration - label: " + groupModel.getLabel());
+
+        // Validate before setting any values
+        for (ConfigModel configModel : groupModel.getGroup()) {
+            ConfigBase config = lookup(configModel.getName());
+            config.validate(configModel);
+        }
+
+        List<ConfigModel> priorConfigList = new LinkedList<>();
+        for (ConfigModel configModel : groupModel.getGroup()) {
+            ConfigBase config = configuration.get(configModel.getName());
+            ConfigModel prior = config.set(configModel);
+            priorConfigList.add(prior);
+        }
+
+        return priorConfigList;
+    }
+
+    public ConfigModel getConfig(String name) {
+        ConfigBase config = lookup(name);
+        return config.get();
+    }
+
+    public List<ConfigModel> getConfigList() {
+        List<ConfigModel> configList = new LinkedList<>();
+        for (ConfigBase config : configuration.values()) {
+            configList.add(config.get());
+        }
+        return configList;
+    }
+
+    public void reset() {
+        for (ConfigBase config : configuration.values()) {
+            config.reset();
+        }
+    }
+
+    // Exposed for use in unit test
+    <T> void addParameter(String name, T value) {
+        if (configuration.containsKey(name)) {
+            throw new DuplicateConfigNameException("Duplicate config name: " + name);
+        }
+
+        ConfigParameter param = new ConfigParameter(name, value);
+        configuration.put(param.getName(), param);
+    }
+
+    public <T> T getParameterValue(String name) {
+        return lookup(name).getCurrentValue();
+    }
+
+    private ConfigBase lookup(String name) {
+        ConfigBase config = configuration.get(name);
+        if (config == null) {
+            throw new ConfigNotFoundException("Unknown configuration name: " + name);
+        }
+        return config;
+    }
+
+    // -- Configuration Setup --
+    public static final String SAM_RETRY_INITIAL_WAIT_SECONDS = "param.sam.retryInitialWaitSeconds";
+    public static final String SAM_RETRY_MAXIMUM_WAIT_SECONDS = "param.sam.retryMaximumWaitSeconds";
+    public static final String SAM_OPERATION_TIMEOUT_SECONDS = "param.sam.operationTimeoutSeconds";
```

Comment:
Could these values be put in an enum instead? Even if the enum were embedded in this class, I think that would give a clearer place to find all the available options, instead of just looking at all static final properties on this class. Also, do we need two strings for a single value : param.sam.retryInitialWaitSeconds and SAM_RETRY_INITIAL_WAIT_SECONDS? Not saying we don't, just checking because I wasn't sure. If you do use an enum and need the separate string, then maybe you could do that with string properties on the enum class?

Reply:
Good point. Initially we were intending for the prefix to denote what type of configuration that string referred to. It seems like `.configType` is handling that now, so I don't see it being as necessary to include in the string. I think unifying the string with the name of the constant variable would improve ergonomics around code search.

Reply:
@marikomedlock @thathert What do you think of the option of putting the enum into the swagger instead of using a string at all?

On the plus side, clients would be able to use the enum and we would get compile-time checking. One the minus side, every time we add a config, we have to update the swagger. And do we want to have 

I can go either way on having the type (e.g. `param`) in the string. If you want a flat name space, we can do that.

"Ergonomics"?

Reply:
Actually, I can't have an enum that works in both a parameter and a property until we convert to swagger v3.

For now, I'll make a separate enum and use it in place of the static strings. I'll make a TODO to remember that it might be nice to move this into swagger

Reply:
I don't like exposing the config parameters in the Swagger, because it seems like a security risk. Maybe that's not a problem if we disable all config settings in production, but not sure if that will always be the case -- maybe we want to leave a few for troubleshooting or something in the future. Anyway, my two sense.